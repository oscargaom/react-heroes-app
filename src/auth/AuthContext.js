import { createContext } from "react";

/*  Al usar createContext este nos facilita información que puede ser 
    compartida con los demás componentes del árbol de componentes */
export const AuthContext = createContext();