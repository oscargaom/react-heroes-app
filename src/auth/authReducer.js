import { types } from "../types/types";

export const authReducer = (state = {}, action) => {
    
    // console.log('=========================')
    // console.log('authReducer');
    // console.log('state');
    // console.log(state);
    // console.log('action');
    // console.log(action);
    // console.log('=========================')
    
    switch (action?.type) {
        case types.login:
            return {
                ...action.payload, 
                logged: true
            }
        case types.logout:
            return {logged: false};
        default:
            return state;
    }
};
