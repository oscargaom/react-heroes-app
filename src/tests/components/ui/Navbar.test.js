import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { Navbar } from '../../../componentes/ui/Navbar';
import { AuthContext } from '../../../auth/AuthContext';
import { authContextLoggedValue, authContextNotLoggedValue } from '../../fixtures/authContextValue';
import { MemoryRouter, Router } from 'react-router-dom';
import { types } from '../../../types/types';

describe('Pruebas en el componente <Navbar />', () => {

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('debe mostrarse correctamente', () => {

        const wrapper = mount(
            <AuthContext.Provider value={authContextLoggedValue}>
                <MemoryRouter>
                    <Navbar />;
                </MemoryRouter>
            </AuthContext.Provider>
        );

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('debe de llamar el logout y usar el history', () => {
        /*  Este Warning nos aparece porque queremos introducir la prop history en el componente 
            <MemoryRouter />: ignores the history prop. To use a custom history, use `import 
            { Router }` instead of `import { MemoryRouter as Router }`.
            Por ello usamos el componente <Router /> y aquì hacemos la inyección de la prop.

            Derivado de este Warning: Failed prop type: Invalid prop `history` of type `function` 
            supplied to `Router`, expected `object`. Agregamos listen al objeto propHistory.
            
            Derivado de este TypeError: history.createHref is not a function. Agregamos 
            createHref: jest.fn() al objeto history.

            Derivado de este TypeError: Cannot read property 'pathname' of undefined. Agregamos
            location: { pathname: '' } al objeto propHistory.

            Derivado de TypeError: history.replace is not a function. Agregamos 
            replace: jest.fn() al objeto propHistory.

        */
        const propHistory = {
            listen: jest.fn(),
            location: {
                pathname: ''
            },
            createHref: jest.fn(),
            replace: jest.fn()
        };

        const wrapper = mount(
            <AuthContext.Provider value={authContextNotLoggedValue}>
                <MemoryRouter>
                    <Router history={propHistory}>
                        <Navbar />
                    </Router>
                </MemoryRouter>
            </AuthContext.Provider>
        );

        // console.log(wrapper.find('button').html())
        wrapper.find('button').props().onClick();

        expect(authContextNotLoggedValue.authDispatch).toHaveBeenCalledTimes(1);
        expect(authContextNotLoggedValue.authDispatch).toHaveBeenCalledWith({
            type: types.logout
        });

        expect(propHistory.replace).toBeCalledTimes(1);
        expect(propHistory.replace).toHaveBeenCalledWith("/login");
    })

})

