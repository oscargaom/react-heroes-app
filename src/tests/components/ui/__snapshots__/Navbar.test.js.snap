// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <Navbar /> debe mostrarse correctamente 1`] = `"<nav class=\\"navbar navbar-expand-sm navbar-dark bg-dark\\"><a class=\\"navbar-brand\\" href=\\"/\\">Asociaciones</a><div class=\\"navbar-collapse\\"><div class=\\"navbar-nav\\"><a class=\\"nav-item nav-link\\" href=\\"/marvel\\">Marvel</a><a class=\\"nav-item nav-link\\" href=\\"/dc\\">DC</a><a class=\\"nav-item nav-link\\" href=\\"/search\\">Search</a></div></div><div class=\\"navbar-collapse collapse w-100 order-3 dual-collapse2\\"><ul class=\\"navbar-nav ml-auto\\"><span class=\\"nav-item nav-link text-info\\"> Manuel </span><button class=\\"nav-item nav-link btn\\">Logout</button></ul></div></nav>;"`;
